tfhe/src/integer/server_key/radix_parallel/div_mod.rs

```diff
@@ -113,134 +114,241 @@ impl ServerKey {
+                    // the one before subtraction
+                    //
+                    // We 'ask' the if_then_else not to clean the message as the first
+                    // operation done on interesting remainder (at the beginning of the loop)
+                    // is a left shift, which will also clean the ciphertext noise, this allows
+                    // to remove 1PBS of latency per loop
+                    let do_clean_message = false;
+                    interesting_remainder = self.unchecked_programmable_if_then_else_parallelized(
+                        &overflowed,
+                        &new_remainder,
+                        &interesting_remainder,
+                        |x| x == 0,
+                        do_clean_message,
+                    );
+                    // As we asked not to clean ciphertext, the degree is not the
+                    // one we want, if we don't fix it, and assert will be hit
+                    // when doing the left shift
+                    for block in &mut interesting_remainder.blocks {
+                        block.degree.0 = self.key.message_modulus.0 - 1;
+                    }
                 },
                 || {
-                    // This is the place where we merge the two cmp blocks
-                    // if it was not done earlier.
-                    let merged_cmp =
-                        is_remainder_greater_or_eq_than_divisor.get_or_insert_with(|| {
-                            self.key.unchecked_apply_lookup_table_bivariate(
-                                &trivial_blocks_are_zero,
-                                &non_trivial_blocks_are_ge.blocks[0],
-                                &merge_two_cmp_lut,
-                            )
-                        });
-                    self.key
-                        .unchecked_scalar_left_shift_assign(merged_cmp, pos_in_block as u8);
-                    self.key
-                        .unchecked_add_assign(&mut quotient.blocks[block_of_bit], merged_cmp);
+                    let did_not_overflow = self.key.unchecked_apply_lookup_table_bivariate(
+                        &overflowed,
+                        &at_least_one_upper_block_is_non_zero,
+                        &merge_overflow_flags_luts[pos_in_block],
+                    );
+
+                    self.key.unchecked_add_assign(
+                        &mut quotient.blocks[block_of_bit],
+                        &did_not_overflow,
+                    );
                 },
             );
+
+            remainder.blocks[..first_trivial_block]
+                .iter_mut()
+                .zip(interesting_remainder.blocks.iter())
+                .for_each(|(remainder_block, new_value)| {
+                    remainder_block.clone_from(new_value);
+                });
         }
 
         // Even though, the quotient does not have any actual carries,
```

Comment:
this does not look true anymore for the remainder, remainder is coming from interestring_remainder itself coming from the unchecked if_then_else without message cleaning it seems to me

Reply:
Yep, depends on if we can keep the fact of not cleaning in the if_then_else